tools/buildsys/src/manifest.rs

```diff
@@ -259,6 +281,89 @@ use std::path::{Path, PathBuf};
 pub struct Error(error::Error);
 type Result<T> = std::result::Result<T, Error>;
 
+#[derive(Debug)]
+pub struct Manifest {
+    graph: PackageGraph,
+    manifest_info: ManifestInfo,
+}
+
+impl Manifest {
+    /// Extract the settings we understand from `Cargo.toml` and construct a dependency graph.
+    pub fn new(manifest: impl AsRef<Path>, cargo_metadata: impl AsRef<Path>) -> Result<Self> {
+        let manifest_info = ManifestInfo::new(manifest)?;
+        let cargo_metadata = cargo_metadata.as_ref();
+        let cargo_metadata_json_contents =
+            fs::read_to_string(cargo_metadata).context(error::CargoMetadataReadSnafu {
+                path: &cargo_metadata,
+            })?;
+        let graph = CargoMetadata::parse_json(cargo_metadata_json_contents)
+            .context(error::CargoMetadataParseSnafu {
+                path: cargo_metadata,
+            })?
+            .build_graph()
+            .context(error::GraphBuildSnafu {
+                path: cargo_metadata,
+            })?;
+        Ok(Self {
+            manifest_info,
+            graph,
+        })
+    }
+
+    /// List all packages that are package dependencies. That is, follow all dependencies in the cargo
+    /// dependency graph that lead to more packages, and do not follow those that involve kits. This
+    /// gives a list of all the packages that are required when we are build a package, or all of the
+    /// packages that should be included when building a kit.
+    pub fn package_dependencies(&self) -> Result<Vec<String>> {
+        let name = self.info().manifest_name();
+        let manifest_type = self.info().build_type()?;
+        let id = find_id(name, &self.graph, manifest_type)
+            .context(error::RootDependencyMissingSnafu { name })?;
+        let ids = [&id];
+        let query = self
+            .graph
+            .query_forward(ids.into_iter())
+            .context(error::CargoPackageQuerySnafuSnafu { id })?;
+        let package_set = query.resolve_with_fn(|_, link| {
+            let to = link.to();
+            is_valid_dep(name, &link) && is_manifest_type(&to, BuildType::Package)
+        });
+        let mut packages: Vec<String> = package_set
+            .packages(DependencyDirection::Forward)
+            .filter_map(|pkg_metadata| filter_map_to_name(name, &pkg_metadata))
+            .collect();
+
+        // Sort so that this function has consistent, dependable output regardless of graph internals.
+        packages.sort();
+        Ok(packages)
```

Comment:
Does this make a difference? Seems like we'd already be getting topologically-sorted output, which should be reasonable consistent. This final sort ends up throwing away the ordering information.

Reply:
> Does this make a difference? Seems like we'd already be getting topologically-sorted output, which should be reasonable consistent. This final sort ends up throwing away the ordering information.

The documentation is silent on order of returned results (though topo-sort seems likely). My most common interaction with this list is "is the package I care about in the list?", whereby alphabetical is more convenient.